aten/src/ATen/native/CPUBlas.h

```diff
@@ -186,4 +192,441 @@ void copy(int64_t n, const float *x, int64_t incx, float *y, int64_t incy);
 void copy(int64_t n, const c10::complex<double> *x, int64_t incx, c10::complex<double> *y, int64_t incy);
 void copy(int64_t n, const c10::complex<float> *x, int64_t incx, c10::complex<float> *y, int64_t incy);
 
-}  // namespace at::native::cpublas
+#if AT_MKLDNN_ENABLED() && (defined(__x86_64__) || (defined(_M_X64) && !defined(_M_ARM64EC)))
+template <typename key_t, typename value_t>
+struct Kernel_Cache {
+  using kstore_t = std::unordered_map<key_t, std::shared_ptr<value_t>>;
+  static inline std::shared_ptr<value_t>&& fetch_or_create(
+      const key_t& key,
+      const std::function<std::shared_ptr<value_t>()>& callback) {
+    auto&& search = get_store().find(key);
+    if (search != get_store().end()) {
+      return std::move(search->second);
```

Comment:
Is it correct to use std::move here? We want to transfer the ownership of the value?

Reply:
This can save copy constructors (some compilers may optimize the return value).